shardtree/src/lib.rs

```diff
@@ -0,0 +1,2851 @@
+                result
+            }
+            Node::Nil => BTreeSet::new(),
+        }
+    }
+
+    /// Prunes the tree by hashing together ephemeral sibling nodes.
+    ///
+    /// `level` must be the level of the root of the node being pruned.
+    pub fn prune(self, level: Level) -> Self {
+        match self {
+            Tree(Node::Parent { ann, left, right }) => Tree::unite(
+                level,
+                ann,
+                left.as_ref().clone().prune(level - 1),
+                right.as_ref().clone().prune(level - 1),
+            ),
+            other => other,
+        }
+    }
+
+    /// Merge two subtrees having the same root address.
+    ///
+    /// The merge operation is checked to be strictly additive and returns an error if merging
+    /// would cause information loss or if a conflict between root hashes occurs at a node. The
+    /// returned error contains the address of the node where such a conflict occurred.
+    pub fn merge_checked(self, root_addr: Address, other: Self) -> Result<Self, Address> {
+        #[allow(clippy::type_complexity)]
+        fn go<H: Hashable + Clone + PartialEq>(
+            addr: Address,
+            t0: PrunableTree<H>,
+            t1: PrunableTree<H>,
+        ) -> Result<PrunableTree<H>, Address> {
+            // Require that any roots the we compute will not be default-filled by picking
+            // a starting valid fill point that is outside the range of leaf positions.
+            let no_default_fill = addr.position_range_end();
+            match (t0, t1) {
+                (Tree(Node::Nil), other) => Ok(other),
+                (other, Tree(Node::Nil)) => Ok(other),
+                (Tree(Node::Leaf { value: vl }), Tree(Node::Leaf { value: vr })) => {
+                    if vl == vr {
+                        Ok(Tree(Node::Leaf { value: vl }))
+                    } else {
+                        Err(addr)
+                    }
+                }
+                (Tree(Node::Leaf { value }), parent) => {
+                    // `parent` is statically known to be a `Node::Parent`
+                    if parent
+                        .root_hash(addr, no_default_fill)
+                        .iter()
+                        .all(|r| r == &value.0)
+                    {
+                        Ok(parent.reannotate_root(Some(Rc::new(value.0))))
+                    } else {
+                        Err(addr)
+                    }
+                }
+                (parent, Tree(Node::Leaf { value })) => {
+                    // `parent` is statically known to be a `Node::Parent`
+                    if parent
+                        .root_hash(addr, no_default_fill)
+                        .iter()
+                        .all(|r| r == &value.0)
+                    {
+                        Ok(parent.reannotate_root(Some(Rc::new(value.0))))
+                    } else {
+                        Err(addr)
+                    }
+                }
```

Comment:
These two cases can be merged together because they behave the same. Also, we can make the pattern match more exhaustive (and the code within more robust) by explicitly pattern matching on the `Node::Parent` case. (The Rust compiler doesn't actually "know" or remember the pattern match holds within the case body, which is annoying.)

```suggestion
               (Tree(Node::Leaf { value }), parent @ Tree(Node::Parent { .. }))
                | (parent @ Tree(Node::Parent { .. }), Tree(Node::Leaf { value })) => {
                    if parent
                        .root_hash(addr, no_default_fill)
                        .iter()
                        .all(|r| r == &value.0)
                    {
                        Ok(parent.reannotate_root(Some(Rc::new(value.0))))
                    } else {
                        Err(addr)
                    }
                }
```